pkg/controller/dnsrecord/actuator.go

```diff
@@ -0,0 +1,145 @@
+	}
+
+	// Update resource status
+	return extensionscontroller.TryUpdateStatus(ctx, retry.DefaultBackoff, a.client, dns, func() error {
+		dns.Status.Zone = &zone
+		return nil
+	})
+}
+
+// Delete deletes the DNSRecord.
+func (a *actuator) Delete(ctx context.Context, dns *extensionsv1alpha1.DNSRecord, cluster *extensionscontroller.Cluster) error {
+	// Create AWS client
+	awsClient, err := aws.NewClientFromDNSSecretRef(ctx, a.client, dns.Spec.SecretRef, dns.Spec.Region)
+	if err != nil {
+		return fmt.Errorf("could not create AWS client: %+v", err)
+	}
+
+	// Determine DNS hosted zone ID
+	zone, err := a.getZone(ctx, dns, awsClient)
+	if err != nil {
+		return err
+	}
+
+	// Delete DNS record
+	recordType := extensionsv1alpha1helper.GetDNSRecordType(dns.Spec.RecordType)
+	ttl := extensionsv1alpha1helper.GetDNSRecordTTL(dns.Spec.TTL)
+	a.logger.Info("Deleting DNS record", "zone", zone, "name", dns.Spec.Name, "type", recordType, "values", dns.Spec.Values)
+	if err := awsClient.DeleteDNSRecord(ctx, zone, dns.Spec.Name, string(recordType), dns.Spec.Values, ttl); err != nil {
+		return fmt.Errorf("could not delete DNS record in zone %s with name %s, type %s, and values %v: %+v", zone, dns.Spec.Name, recordType, dns.Spec.Values, err)
+	}
+
+	// Update resource status
+	return nil
+}
+
+// Restore restores the DNSRecord.
+func (a *actuator) Restore(ctx context.Context, dns *extensionsv1alpha1.DNSRecord, cluster *extensionscontroller.Cluster) error {
+	return a.Reconcile(ctx, dns, cluster)
+}
+
+// Migrate migrates the DNSRecord.
+func (a *actuator) Migrate(ctx context.Context, dns *extensionsv1alpha1.DNSRecord, cluster *extensionscontroller.Cluster) error {
+	return nil
+}
+
+func (a *actuator) getZone(ctx context.Context, dns *extensionsv1alpha1.DNSRecord, awsClient awsclient.Interface) (string, error) {
+	switch {
+	case dns.Status.Zone != nil && *dns.Status.Zone != "":
+		return *dns.Status.Zone, nil
+	case dns.Spec.Zone != nil && *dns.Spec.Zone != "":
+		return *dns.Spec.Zone, nil
+	default:
+		// The zone is not specified in the resource status or spec. Try to determine the zone by
+		// getting all hosted zones of the account and searching for the longest zone name that is a suffix of dns.spec.Name
+		zones, err := awsClient.GetDNSHostedZones(ctx)
+		if err != nil {
+			return "", fmt.Errorf("could not get DNS hosted zones: %+v", err)
+		}
+		a.logger.Info("Got DNS hosted zones", "zones", zones)
+		zone := findZoneForName(zones, dns.Spec.Name)
+		if zone == "" {
+			return "", fmt.Errorf("could not find DNS hosted zone for name %s", dns.Spec.Name)
+		}
+		return zone, nil
+	}
+}
+
+func findZoneForName(zones map[string]string, name string) string {
+	longestZoneName, result := "", ""
+	for zoneName, zoneId := range zones {
+		if strings.HasSuffix(name, zoneName) && len(zoneName) > len(longestZoneName) {
+			longestZoneName, result = zoneName, zoneId
+		}
```

Comment:
It will not (`strings.HasSuffix("d.c.b.a", "ccc.b.a") == false`). You probably mean "d.cccc.b.a" must not match "ccc.b.a", in which case I agree. Will use the propose matching function, thanks!

Reply:
Done.